cfgrammar/src/lib/span.rs

```diff
@@ -44,72 +44,98 @@ impl Span {
         }
     }
+}
 
-    fn offset_to_line_col(&mut self, off: usize) -> (usize, usize) {
-        if self.high_water < off {
-            self.newlines.extend(
-                // assumes off is_char_boundary would return true.
-                self.src[self.high_water..off]
-                    .char_indices()
-                    .filter_map(|c| match c {
-                        (offset, '\n') | (offset, '\r') => Some(offset + 1),
-                        _ => None,
-                    }),
-            );
-
-            self.high_water = off;
-        }
+impl NewlineToLineColCache {
+    /// Feed more input into the cache, calculating newlines data from it.
+    /// The `src` string given is treated as though it is contiguous with previous a call to `feed`.
+    /// subsequent logical lines, and logical bytes reflecting the sum of total of offsets which
+    /// have been fed.
+    pub fn feed(&mut self, src: &str) {
+        let start_pos = self.newlines[self.newlines.len() - 1] + self.trailing_bytes;
+        self.newlines
+            .extend(src.char_indices().filter_map(|c| match c {
+                (offset, '\n') | (offset, '\r') => {
+                    self.trailing_bytes = 0;
+                    Some(start_pos + offset + 1)
+                }
+                _ => {
+                    self.trailing_bytes += 1;
+                    None
+                }
+            }));
+    }
 
-        if off == self.src.len() {
-            let line_off = self.newlines.last().unwrap();
-            return (
-                self.newlines.len(),
-                self.src[*line_off..].chars().count() + 1,
-            );
+    /// Convert a logical byte offset in the input to a logical line number. Returns None if
+    /// the byte offset exceeds the known input length.
+    pub fn byte_to_line_num(&self, byte: usize) -> Option<usize> {
+        let last_newline = self.newlines.last().unwrap();
+        let last_byte = last_newline + self.trailing_bytes;
+        if byte < last_byte && byte > *last_newline {
+            Some(self.newlines.len())
+        } else {
+            let (line_m1, _) = self
+                .newlines
+                .iter()
+                .enumerate()
+                .rev()
+                .find(|&(_, &line_off)| line_off <= byte)
+                .unwrap();
+            Some(line_m1 + 1)
         }
-
-        let (line_m1, &line_off) = self
-            .newlines
-            .iter()
-            .enumerate()
-            .rev()
-            .find(|&(_, &line_off)| line_off <= off)
-            .unwrap();
-
-        let c_off = self.src[line_off..]
-            .char_indices()
-            .position(|(c_off, _)| c_off == off - line_off)
-            .unwrap();
-        (line_m1 + 1, c_off + 1)
     }
 
-    pub fn start_to_line_col(&mut self, span: Span) -> (usize, usize) {
-        self.offset_to_line_col(span.start())
+    /// Convert a logical line number into a logical byte offset. Returns None
```

Comment:
I'm not sure what a "logical byte" is.

Reply:
Yeah, i wasn't certain if I should introduce that or not, but I meant it as indicating that it was a byte relative to the concatenation of all strings fed.  I could try and improve that or remove it wtdyt?

Reply:
I think "logical" is misleading here, because it's just the "byte offset since the start of the input". The reason we have "logical lines" is because they don't match up 1:1 to bytes.